saleor/order/actions.py

```diff
@@ -1361,37 +1380,51 @@ def _process_refund(
     refund_amount = _calculate_refund_amount(
         order_lines_to_refund, fulfillment_lines_to_refund, lines_to_refund
     )
+
+    # Happens when neither payments_to_refund nor amount_to_refund were specified.
     if amount is None:
         amount = refund_amount
         # we take into consideration the shipping costs only when amount is not
         # provided.
         if refund_shipping_costs:
             amount += order.shipping_price_gross_amount
+        # At this point there can be only one payment.
+        payments[0]["amount"] = min(payments[0]["payment"].captured_amount, amount)
+
     if amount:
-        amount = min(payment.captured_amount, amount)
-        try:
-            gateway.refund(
-                payment, manager, amount=amount, channel_slug=order.channel.slug
-            )
-        except PaymentError:
-            raise ValidationError(
-                "The refund operation is not available yet.",
-                code=OrderErrorCode.CANNOT_REFUND.value,
-            )
-        transaction.on_commit(
-            lambda: events.payment_refunded_event(
-                order=order,
-                user=user,
-                app=app,
-                amount=amount,  # type: ignore
-                payment=payment,
+        for item in payments:
+            try:
```

Comment:
Assuming that you have 3 payments. 
1st will be processed with success, the second also, then the 3rd one will raise an error, which will raise a ValidationError, in that case I think, that we will not have any events about success payment refund (which are called in transaction.on_commit.

Reply:
Done